File: c/src/runner.c
#include "bench.h"
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <sys/stat.h>
#include <errno.h>

uint64_t bench_now_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ull + (uint64_t)ts.tv_nsec;
}

static int mkdir_p(const char *path) {
    char tmp[512];
    snprintf(tmp, sizeof(tmp), "%s", path);
    size_t len = strlen(tmp);
    if (len == 0) return -1;
    for (size_t i = 1; i < len; i++) {
        if (tmp[i] == '/') {
            tmp[i] = 0;
            if (mkdir(tmp, 0755) && errno != EEXIST) return -1;
            tmp[i] = '/';
        }
    }
    if (mkdir(tmp, 0755) && errno != EEXIST) return -1;
    return 0;
}

static bool name_match(const char *name, const char *filter) {
    if (!filter || !filter[0]) return true;
    /* case-insensitive substring */
    size_t n = strlen(name), f = strlen(filter);
    for (size_t i = 0; i + f <= n; i++) {
        size_t j = 0;
        for (; j < f; j++) {
            char a = name[i + j], b = filter[j];
            if (a >= 'A' && a <= 'Z') a += 32;
            if (b >= 'A' && b <= 'Z') b += 32;
            if (a != b) break;
        }
        if (j == f) return true;
    }
    return false;
}

int run_benchmarks(int repetitions, const char *ser_filter, const char *data_filter,
                   const char *log_dir) {
    serializer_t sers[BENCH_MAX_SERIALIZERS];
    int ser_count = 0;
    register_all_serializers(sers, &ser_count);

    test_fixture_t fixtures[TD_COUNT];
    data_init_all(fixtures, TD_COUNT, 42);

    char path[512];
    snprintf(path, sizeof(path), "%s", log_dir ? log_dir : "logs/c");
    mkdir_p(path);

    // Timestamped result file. BENCHMARK_TS set by orchestrator for consistent naming.
    const char *ts = getenv("BENCHMARK_TS");
    char log_path[600];
    if (ts && strlen(ts) > 8) {
        snprintf(log_path, sizeof(log_path), "%s/%s.csv", path, ts);
    } else {
        // local fallback
        snprintf(log_path, sizeof(log_path), "%s/local-%ld.csv", path, (long)time(NULL));
    }
    csv_logger_t *log = csv_logger_create(log_path);
    if (!log) {
        fprintf(stderr, "Cannot create log %s\n", log_path);
        return 1;
    }

    static uint8_t buf[4 * 1024 * 1024];
    const char *modes[] = { "bytes", "stream" };

    printf("[PROGRESS] C benchmark: %d serializers, %d data types, %d reps\n",
           ser_count, TD_COUNT, repetitions);

    for (int di = 0; di < TD_COUNT; di++) {
        test_fixture_t *fx = &fixtures[di];
        if (!name_match(fx->name, data_filter)) continue;
        printf("[PROGRESS] Testing Data: %s\n", fx->name);

        for (int si = 0; si < ser_count; si++) {
            serializer_t *S = &sers[si];
            if (!name_match(S->name, ser_filter)) continue;
            if (S->supports && !S->supports(fx->kind)) continue;
            if (S->prepare && S->prepare(fx->kind, fx) != 0) continue;

            for (int mi = 0; mi < 2; mi++) {
                const char *mode = modes[mi];
                int had_error = 0;
                /* Log every successful rep including r==0 (warmup). Analysis drops warmup later. */
                for (int r = 0; r < repetitions; r++) {
                    size_t out_len = 0;
                    test_fixture_t out_fx;
                    memset(&out_fx, 0, sizeof(out_fx));
                    out_fx.kind = fx->kind;

                    uint64_t t0 = bench_now_ns();
                    int rc = S->serialize(fx, buf, sizeof(buf), &out_len);
                    if (rc == 0 && mode[0] == 's')
                        rc = bench_stream_write_all(buf, out_len);
                    uint64_t t1 = bench_now_ns();
                    if (rc != 0) {
                        if (!had_error) {
                            fprintf(stderr, "[ERROR] %s / %s / %s: serialize failed\n",
                                    S->name, fx->name, mode);
                            had_error = 1;
                        }
                        continue;
                    }
                    if (mode[0] == 's') {
                        if (bench_stream_read_all(buf, sizeof(buf), out_len) != 0) {
                            if (!had_error) {
                                fprintf(stderr, "[ERROR] %s / %s / %s: stream read failed\n",
                                        S->name, fx->name, mode);
                                had_error = 1;
                            }
                            continue;
                        }
                    }
                    rc = S->deserialize(buf, out_len, &out_fx, fx->kind);
                    uint64_t t2 = bench_now_ns();
                    if (rc != 0) {
                        if (!had_error) {
                            fprintf(stderr, "[ERROR] %s / %s / %s: deserialize failed\n",
                                    S->name, fx->name, mode);
                            had_error = 1;
                        }
                        continue;
                    }
                    bool ok = S->fidelity ? S->fidelity(fx, &out_fx) : true;
                    if (!ok) {
                        if (!had_error) {
                            fprintf(stderr, "[ERROR] %s / %s / %s: fidelity failed\n",
                                    S->name, fx->name, mode);
                            had_error = 1;
                        }
                        continue;
                    }
                    if (!had_error) {
                        const char *sm = (mode && strcmp(mode, "stream") == 0) ? "adapted" : "";
                        size_t gz = 0, zs = 0;
                        bench_compress_sizes(buf, out_len, &gz, &zs);
                        csv_logger_write(log, mode, fx->name, repetitions, r, S->name,
                                         t1 - t0, t2 - t1, out_len, 1.0, S->version,
                                         1, "", sm, -1, -1, gz, zs);
                    }
                }
            }
        }
    }
    csv_logger_close(log);
    printf("[PROGRESS] Complete. Results: %s\n", log_path);
    return 0;
}

File: php/src/Runner.php
<?php

declare(strict_types=1);

namespace Benchmark;

use Benchmark\Serializers\Registry;
use Benchmark\Serializers\Serializer;

require dirname(__DIR__) . '/vendor/autoload.php';

$repetitions = (int) ($argv[1] ?? 10);
$serFilter = (string) ($argv[2] ?? '');
$dataFilter = strtolower((string) ($argv[3] ?? ''));

$projectRoot = dirname(__DIR__, 2);
$logDirEnv = getenv('LOG_DIR') ?: ($projectRoot . '/logs');
$logDir = str_ends_with(rtrim($logDirEnv, '/'), 'php') ? $logDirEnv : ($logDirEnv . '/php');
if (!is_dir($logDir) && !mkdir($logDir, 0777, true) && !is_dir($logDir)) {
    fwrite(STDERR, "cannot create $logDir\n");
    exit(1);
}

$ts = getenv('BENCHMARK_TS') ?: date('Y-m-d-His');
$logPath = $logDir . '/' . $ts . '.csv';
$errPath = $logDir . '/' . $ts . '.errors.csv';

$header = 'Language,StringOrStream,TestDataName,Repetitions,RepetitionIndex,SerializerName,SerializerVersion,TimeSer,TimeDeser,Size,TimeSerAndDeser,OpPerSecSer,OpPerSecDeser,OpPerSecSerAndDeser,MemoryPeakBytes,FidelityScore,NativeKind,StreamMode,DataTypeInstanceCount,TypeConfigHash,RunOrder,SchedulePosition,SizeGzip,SizeZstd' . "\n";
$lines = [$header];
$errors = ['TestDataName,SerializerName,StringOrStream,Repetition,ErrorText' . "\n"];

$resolved = resolveRunConfig($projectRoot);
$seed = (int) (getenv('BENCHMARK_SEED') ?: ($resolved['seed'] ?? 42));
$cells = $resolved['cells'] ?? [];
if ($dataFilter !== '') {
    $cells = array_values(array_filter(
        $cells,
        static fn ($c) => str_contains(strtolower((string) ($c['type_id'] ?? '')), $dataFilter),
    ));
}
$modesRaw = $resolved['execution']['io_modes'] ?? ['bytes'];
$modes = [];
foreach ($modesRaw as $m) {
    $m = strtolower((string) $m);
    if ($m === 'bytes' || $m === 'stream') {
        $modes[$m] = $m;
    }
}
$modes = array_values($modes);
if ($modes === []) {
    $modes = ['bytes'];
}

$serializers = Registry::select($serFilter);
$strategy = Schedule::strategy();
$recordRO = Schedule::recordRunOrder();

fwrite(STDERR, sprintf(
    "[PROGRESS] PHP Data Model v2: %d serializers, %d cells, %d reps, modes=[%s] schedule=%s\n",
    count($serializers),
    count($cells),
    $repetitions,
    implode(',', $modes),
    $strategy,
));

$runOrder = 0;
foreach ($cells as $cell) {
    $n = (int) ($cell['data_type_instance_count'] ?? 1);
    $typeId = (string) $cell['type_id'];
    $cfg = $cell['type_config'] ?? [];
    $hash = (string) ($cell['type_config_hash'] ?? '');
    $value = $n === 1
        ? DataV2::makeOne($typeId, $cfg, $seed, 0)
        : DataV2::instances($typeId, $cfg, $seed, $n);
    fwrite(STDERR, "[PROGRESS] Testing Data: {$typeId} (N={$n})\n");

    $ready = [];
    foreach ($serializers as $ser) {
        try {
            $ser->prepare($value);
            $gz = 0;
            $zs = 0;
            try {
                $sample = $ser->serializeBytes($value);
                $gz = strlen(gzencode($sample, 6) ?: '');
            } catch (\Throwable) {
                $gz = 0;
            }
            $ready[$ser->name()] = ['ser' => $ser, 'gz' => $gz, 'zs' => $zs];
        } catch (\Throwable $e) {
            fwrite(STDERR, sprintf("[ERROR] prepare %s / %s: %s\n", $ser->name(), $typeId, $e->getMessage()));
            $errors[] = csvRow([$typeId, $ser->name(), 'prepare', '0', $e->getMessage()]);
        }
    }

    foreach ($modes as $mode) {
        for ($rep = 0; $rep < $repetitions; $rep++) {
            $names = array_keys($ready);
            if ($strategy !== 'none') {
                $names = Schedule::shuffle($names, $seed, $typeId, $n, $hash, $mode, $rep);
            }
            $pos = 0;
            foreach ($names as $name) {
                /** @var Serializer $ser */
                $ser = $ready[$name]['ser'];
                $pos++;
                $runOrder++;
                try {
                    if ($mode === 'bytes') {
                        $t0 = hrtime(true);
                        $bytes = $ser->serializeBytes($value);
                        $t1 = hrtime(true);
                        $back = $ser->deserializeBytes($bytes);
                        $t2 = hrtime(true);
                        $size = strlen($bytes);
                    } else {
                        $buf = fopen('php://memory', 'r+b');
                        $t0 = hrtime(true);
                        $size = $ser->serializeStream($value, $buf);
                        $t1 = hrtime(true);
                        rewind($buf);
                        $back = $ser->deserializeStream($buf);
                        $t2 = hrtime(true);
                        fclose($buf);
                    }
                    $serNs = $t1 - $t0;
                    $deserNs = $t2 - $t1;
                    $tot = $serNs + $deserNs;
                    $ok = fidelity($value, $back) ? 1.0 : 0.0;
                    if ($ok < 1.0) {
                        $errors[] = csvRow([$typeId, $name, $mode, (string) $rep, 'fidelity']);
                    }
                    $lines[] = csvRow([
                        'php',
                        $mode,
                        $typeId,
                        (string) $repetitions,
                        (string) $rep,
                        $name,
                        $ser->version(),
                        (string) $serNs,
                        (string) $deserNs,
                        (string) $size,
                        (string) $tot,
                        $serNs > 0 ? formatOps($serNs) : '',
                        $deserNs > 0 ? formatOps($deserNs) : '',
                        $tot > 0 ? formatOps($tot) : '',
                        '0',
                        (string) $ok,
                        'message',
                        $ser->streamMode(),
                        (string) $n,
                        $hash,
                        $recordRO ? (string) $runOrder : '',
                        (string) $pos,
                        (string) $ready[$name]['gz'],
                        (string) $ready[$name]['zs'],
                    ]);
                } catch (\Throwable $e) {
                    $errors[] = csvRow([$typeId, $name, $mode, (string) $rep, $e->getMessage()]);
                }
            }
        }
    }
}

file_put_contents($logPath, implode('', $lines));
if (count($errors) > 1) {
    file_put_contents($errPath, implode('', $errors));
}
fwrite(STDERR, "[PROGRESS] Complete. Results: {$logPath}\n");

function resolveRunConfig(string $projectRoot): array
{
    $runCfg = getenv('BENCHMARK_RUN_CONFIG') ?: ($projectRoot . '/config/library/default.yaml');
    if (!str_starts_with($runCfg, '/')) {
        $fromRepo = $projectRoot . '/' . ltrim($runCfg, './');
        $runCfg = is_file($fromRepo) ? $fromRepo : $runCfg;
    }
    $seed = getenv('BENCHMARK_SEED') ?: '42';
    $script = $projectRoot . '/scripts/resolve_run_config.py';
    $cmd = [
        'python3',
        $script,
        $runCfg,
        '--seed',
        $seed,
    ];
    $env = [];
    foreach (getenv() as $k => $v) {
        if (is_string($v)) {
            $env[$k] = $v;
        }
    }
    $prev = $env['PYTHONPATH'] ?? '';
    $env['PYTHONPATH'] = $projectRoot . '/analysis/src' . ($prev !== '' ? ':' . $prev : '');
    $proc = proc_open($cmd, [1 => ['pipe', 'w'], 2 => ['pipe', 'w']], $pipes, $projectRoot, $env);
    if (!is_resource($proc)) {
        throw new \RuntimeException('resolve_run_config spawn failed');
    }
    $out = stream_get_contents($pipes[1]);
    $err = stream_get_contents($pipes[2]);
    fclose($pipes[1]);
    fclose($pipes[2]);
    $code = proc_close($proc);
    if ($code !== 0) {
        throw new \RuntimeException('resolve_run_config failed: ' . $err . $out);
    }
    return json_decode($out, true, 512, JSON_THROW_ON_ERROR);
}

function fidelity(mixed $a, mixed $b): bool
{
    if (is_array($a) && is_array($b)) {
        if (count($a) !== count($b)) {
            return false;
        }
        foreach ($a as $k => $x) {
            if (!array_key_exists($k, $b) || !fidelity($x, $b[$k])) {
                return false;
            }
        }
        return true;
    }
    if (is_bool($a)) {
        if (is_bool($b)) {
            return $a === $b;
        }
        return $a === filter_var($b, FILTER_VALIDATE_BOOLEAN);
    }
    if (is_int($a) && is_numeric($b)) {
        return $a === (int) $b;
    }
    if (is_float($a) || is_float($b) || (is_numeric($a) && is_numeric($b) && (is_string($a) || is_string($b)))) {
        return abs((float) $a - (float) $b) < 1e-8;
    }
    return $a === $b;
}

function formatOps(int $ns): string
{
    return number_format(1_000_000_000 / $ns, 6, '.', '');
}

function csvRow(array $cols): string
{
    $esc = [];
    foreach ($cols as $c) {
        $s = (string) $c;
        if (str_contains($s, ',') || str_contains($s, '"') || str_contains($s, "\n")) {
            $s = '"' . str_replace('"', '""', $s) . '"';
        }
        $esc[] = $s;
    }
    return implode(',', $esc) . "\n";
}
